pkg/reconcilermanager/controllers/garbage_collector.go

```diff
@@ -233,8 +232,32 @@ func (r *reconcilerBase) deleteDeployment(ctx context.Context, reconcilerRef typ
 	return r.cleanup(ctx, d)
 }
 
-func (r *RootSyncReconciler) deleteClusterRoleBinding(ctx context.Context, name string) error {
-	crb := &rbacv1.ClusterRoleBinding{ObjectMeta: metav1.ObjectMeta{Name: name}}
-
-	return r.cleanup(ctx, crb)
+func (r *RootSyncReconciler) deleteSharedClusterRoleBinding(ctx context.Context, name string, reconcilerRef types.NamespacedName) error {
+	crbKey := client.ObjectKey{Name: name}
+	// Update the CRB to delete the subject for the deleted RootSync's reconciler
+	crb := &rbacv1.ClusterRoleBinding{}
+	if err := r.client.Get(ctx, crbKey, crb); err != nil {
+		if apierrors.IsNotFound(err) {
+			// already deleted
+			r.logger(ctx).Info("Managed object already deleted",
+				logFieldObjectRef, crbKey.String(),
+				logFieldObjectKind, "ClusterRoleBinding")
+			return nil
+		}
+		return NewObjectOperationErrorWithKey(err, crb, OperationGet, crbKey)
+	}
+	count := len(crb.Subjects)
+	crb.Subjects = removeSubject(crb.Subjects, r.serviceAccountSubject(reconcilerRef))
+	if count == len(crb.Subjects) {
+		// No change
+		return nil
+	}
+	if len(crb.Subjects) == 0 {
```

Comment:
If `count` was zero when it's defined (L249), this line won't be executed.

Reply:
To be clear the behavior is actually unchanged from before, although the diff is a bit misleading. The original function for reference: https://github.com/GoogleContainerTools/kpt-config-sync/blob/v1.16/pkg/reconcilermanager/controllers/garbage_collector.go#L235

I agree with you though that we should switch the order so that the CRB always gets garbage collected when the subjects are empty. I'll update this